### PR TITLE
Add `experimental_canvas_size_limits` argument to points

### PR DESCRIPTION
**Pull request by brisvag**
# Description
*Remaking this PR cause somehow I had managed to make the other from a branch on the main repo...*

This PR adds the ~`fixed_canvas_size`~ argument to the `Points` layer. When active, point sizes are no longer defined in data space, but canvas space, which means they will stay fixed when zooming.

This was split out of #3430.

EDIT: this now instead implements `(experimental_)canvas_size_limits` (see below for discussions on why), which works similarly to contrast limits: it's a tuple that defines low and high limits for the canvas size of the points. Everything happens in the shader, which means it's fast, but also [creates problems with `get_value`](https://github.com/napari/napari/pull/3734#issuecomment-987889759). For this reason, this is for now left as `experimental_`.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
- #429

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [x] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by brisvag** [comment]
So we don't lose track of it:

> On a different point: importantly, this reminded me that there was an issue we had to address when implementing this feature. [We talked about it in this PR](https://github.com/napari/napari/pull/3508#discussion_r746526151), but in short: we need `get_value` to behave differently when `fixed_canvas_size` is active.

File: napari/layers/points/points.py
Comment (brisvag) [inline napari/layers/points/points.py]: This is where we should calculate distances differently. @alisterburt @kevinyamauchi , I'm not sure where to get the right transforms. I see we have data_to_world and friends, but do we expose any canvas data?
Alternatively, how would you approach this?

Comment (kevinyamauchi) [reply napari/layers/points/points.py]: I think it depends on whether we want to support perspective or not. For orthographic projection, we can use the camera zoom to calculate the canvas pixel size in data units and use that. The tricky part is getting access to the scale layer.

See how the calculation is done for the axis visual: https://github.com/napari/napari/blob/fc133a627dc92aff0297a010d7e2e4f5d3a82092/napari/_vispy/overlays/axes.py#L328-L343

Comment (brisvag) [reply napari/layers/points/points.py]: But we don't have access to viewer and camera from the layer, and we probably shouldn't... This might be a problem :P

Comment (kevinyamauchi) [reply napari/layers/points/points.py]: Yeah, agreed. This is one of the reasons we have been punting on interactivity with perspective. I would be okay with proceeding without for now seeing as perspective breaks interactivity for all other layers.

Comment (kevinyamauchi) [reply napari/layers/points/points.py]: I can review more carefully tomorrow!

**Comment by sofroniewn** [comment]
@brisvag @kevinyamauchi should we consider a min/ max size here so that this PR can address https://github.com/napari/napari/issues/429 or do we think fixed size is sufficient. It would be nice if this PR could close that. cc @berl

**Comment by brisvag** [comment]
To get a min/max size we probably need the same machinery that we need anyways to allow `get_value` to work properly (see above). I think we'll need to sit down and figure this out some time @kevinyamauchi !

**Comment by kevinyamauchi** [comment]
> To get a min/max size we probably need the same machinery that we need anyways to allow get_value to work properly (see above). I think we'll need to sit down and figure this out some time @kevinyamauchi !

Agreed!

File: napari/_vispy/layers/points.py
Comment (kevinyamauchi) [inline napari/_vispy/layers/points.py]: I think a comment here would be nice.

File: napari/_vispy/visuals/points.py
Comment (kevinyamauchi) [inline napari/_vispy/visuals/points.py]: I think it would be nice to add a short docstring explaining the scaling property.

File: napari/layers/points/points.py
Comment (kevinyamauchi) [inline napari/layers/points/points.py]:
```suggestion
        If set to true, the point size is in canvas units and thus will not change size on the screen
        when zooming. If set to False, the point size is set in data units. The default value is False.
```

File: napari/layers/points/points.py
Comment (kevinyamauchi) [inline napari/layers/points/points.py]:
```suggestion
        If set to true, the point size is in canvas units and thus will not change size on the screen
        when zooming. If set to False, the point size is set in data units
```

**Review by kevinyamauchi — COMMENTED** [review]
Thanks, @brisvag! I think this looks good. I tested it and it seems to be working. I left some minor comments below. I also have a couple of questions/suggestions:

- I think we should add a simple test where the `fixed_canvas_size` is set on the model and we check if the node is updated properly. See the tests [here](https://github.com/napari/napari/blob/main/napari/_vispy/_tests/test_vispy_points_layer.py)
- Should we add a GUI element to the layer controls (e.g., a checkbox)? I would be okay with this being a follow-up PR.

File: napari/layers/points/points.py
Comment (brisvag) [reply napari/layers/points/points.py]: To be clear: this is not an issue only with perspective. Even with fov=0, without additional changes and with `fixed_canvas_size` active we get basically the same issue as #3610.

**Comment by brisvag** [comment]
Addedd comments, docstrings and test. As for the gui: Maybe before making it so "public" we need to address the issue with selection/highlighting? It depends how bad we consider that to be.

**Comment by jni** [comment]
@brisvag @kevinyamauchi what do you think of my suggestion [here](https://github.com/napari/napari/pull/3430#issuecomment-984215486) to change the API to `minimum_canvas_size` and `maximum_canvas_size`? These would default to 0 (or some other small value) and np.inf, and to set `fixed_canvas_size` you would set both.

To be clear, I'm happy to merge this as is but I'd like to have a proper discussion about it. In particular, my own usage preference would be to only use minimum, to make sure that points are visible at any zoom level. I think that would be really nice and would resolve a recurring issue I have which is that 10 is too big as a default point size, but 1 is hard to see in many datasets.

To me, that is a much more useful API. It's the same motivation as #429 (by my reading), whereas fixed canvas size doesn't really resolve that issue. Does anyone have a real world use case for *really* fixed canvas size?

**Comment by brisvag** [comment]
@kevinyamauchi  and I had a chat, and this is what we came up with:

1. The real difficulty in this PR lies on making `get_value` work when sizes are not fixed in data space. This is currently impossible because the layer (and therefore `get_value`) is unaware of the camera and the canvas, which are needed to calculate the current screen-space size. Solutions:
    - pass from `qt_viewer` to the mouse event a `world2canvas` transform, so we can use it in the `get_value` calculation
    - make the layer aware of the canvas/camera and have a `data2canvas` transform itself

   The first option seems better, though we're not happy with how big the event baggage is becoming.

2. We could merge this *without* any of the above, if people are OK with `get_value` (and thus the ability to select points, for example) to be broken when these options are enabled. @kevinyamauchi and I would say no to this, as that would be quite disruptive in terms of UX.

3. the min/max sizes should be trivial to implement on the shader level (just clamp `gl_PointSize` in the vertex shader), so that's easy to add. Both should be about the same to implement based on what we do above.

> Does anyone have a real world use case for really fixed canvas size?

I actually think `fixed_canvas_size` makes more sense for the cases where you need it. My reasoning is: if absolute point sizes (in data space) are important data (i.e: atom sizes in a molecule) you never want them to change absolute size, no matter how much you zoom in/out. On the other hand, if sizes only have *relative* meaning (i.e: a plot where point sizes represent the population of cities), then you might as well set a single value to True, rather than 2 floats that are arbitrarily clamping the range in which you want absolute sizes.

This is not to say that these option shuldn't exist, just that I personally don't see where I'd use `min/max` rathern than just `fixed`.

Ultimately, I think we can/should support both `min/max` and `fixed`, (with fixed having priority?), since they ~are~ should be equally hard/easy to implement. The api will expand even more, though.

**Comment by psobolewskiPhD** [comment]
> 3\. the min/max sizes should be trivial to implement on the shader level (just clamp `gl_PointSize` in the vertex shader), so that's easy to add. Both should be about the same to implement based on what we do above.

Am I correct in thinking that this `clamping` would essentially fix the macOS point layer issues? https://github.com/napari/napari/issues/3610
Points would appear the correct size, as limited by `gl_PointSize`, so there would be no selection ambiguity?

**Comment by brisvag** [comment]
Only if we address number `1`, and add a system to keep track of `MAX_GL_POINT_SIZE` or whatever it's called. Without doing `1`, this PR would essentially cause exactly the same issue you're having in #3610 for anyone who uses `fixed_canvas_size` or `min/max_canvas_size`.

**Comment by psobolewskiPhD** [comment]
Ah, bummer! I saw `trivial` and got excited 🤣

**Comment by sofroniewn** [comment]
> The real difficulty in this PR lies on making get_value work when sizes are not fixed in data space. This is currently impossible because the layer (and therefore get_value) is unaware of the camera and the canvas, which are needed to calculate the current screen-space size. Solutions:

oof, yeah, this is maybe a little hairy. i don't love the idea that every get value call or every layer would now get this canvas information just to meet this use case in points, which is ultimately quite different that our other usages (in the sense that the data cannot be thought of as data in a physical world anymore, but is now something not physical). I can't really see a way around it though. 

> pass from qt_viewer to the mouse event a world2canvas transform, so we can use it in the get_value calculation
> ....
> The first option seems better, though we're not happy with how big the event baggage is becoming.

totally agreed here, i prefer this option too - but i don't love it either. lots of complexity here. i wonder if @andy-sweet might have a good idea?

**Comment by berl** [comment]
I like 2 better than 1. How critical is the `get_value` functionality when the point size is canvas/camera dependent? @brisvag  why would selecting points not work without it?   To me it seems like unexpected behavior in setting the value could be more painful for users- you select your points and go to bigger them only to see nothing happen. 

@sofroniewn  the rendered marker size still has a well-defined physical meaning at any point in time (as long as we're not in 3d rendering or talking about 3d markers), it's just no longer connected to the value it was given in the layer.  

I look forward to seeing how this gets worked out!

**Comment by jni** [comment]
@berl the issue is that to select a point you click "on it", and to know whether a click falls on a point, you need to know the size of the point on the canvas — which the layer (which handles these clicks) doesn't know anymore.

**`@everyone`**, I suspect that the way to fix this is to update our transforms system together with our slicing. A layer should definitely not know about canvas coordinates, but a *sliced layer* can. I'll write an issue about how I see transforms going (been thinking about this for a while in the context of both plugins and geodata), and I suspect @perlman @nclack will have notes to add.

**Comment by brisvag** [comment]
Yeah, that sounds like a good plan. I don't think this is a high priority feature, so I'm ok with putting in on hold. Alternatively, we could expose it as `experimental_` with a warning in the docs that it breaks selection, but at least it would allow people to work around the issue if needed.

**Comment by sofroniewn** [comment]
> I suspect that the way to fix this is to update our transforms system together with our slicing. A layer should definitely not know about canvas coordinates, but a sliced layer can. I'll write an issue about how I see transforms going

Sounds good, makes sense - i think we'll be on same page

>  Alternatively, we could expose it as experimental_ with a warning in the docs that it breaks selection, but at least it would allow people to work around the issue if needed.

I like that. @jni?

**Comment by jni** [comment]
@brisvag 

> My reasoning is: if absolute point sizes (in data space) are important data (i.e: atom sizes in a molecule) you never want them to change absolute size, no matter how much you zoom in/out. On the other hand, if sizes only have relative meaning (i.e: a plot where point sizes represent the population of cities), then you might as well set a single value to True, rather than 2 floats that are arbitrarily clamping the range in which you want absolute sizes.

I don't think this analysis is complete. The analogy is with intensity images and contrast limits: you are displaying the intensity of an image, clipping things below and above an area of interest. Making the contrast limits narrower does not mean that you don't care about any variability or relative intensity, it means that you are focusing on a particular range. In the case of points, there is absolutely a use case: point sizes have some relative meaning, and I want to see that meaning, *but not so much that points become invisible (too small) or take up the entire screen*. I think actually min/max canvas size is useful *generally*, ie it might be good to turn it on by default.

> I'll write an issue about how I see transforms going

I suspect most people have seen this already but linking here for completeness: #3783 

> Alternatively, we could expose it as experimental_ with a warning in the docs that it breaks selection

Does this work for you @berl?

I still would like to do min/max vs fixed size, for the reasons above, even if it's experimental. Ultimately I'm agnostic about `experimental_` stuff: if people want it in, fine, it's easy to revert/remove it if it causes too many issues. But adding the option to the UI is harder, though.

Transforms could take a long while to land in a happy place though, so if someone is really pining for this, we can press on.

**Comment by brisvag** [comment]
> I don't think this analysis is complete. The analogy is with intensity images and contrast limits: you are displaying the intensity of an image, clipping things below and above an area of interest. Making the contrast limits narrower does not mean that you don't care about any variability or relative intensity, it means that you are focusing on a particular range. In the case of points, there is absolutely a use case: point sizes have some relative meaning, and I want to see that meaning, _but not so much that points become invisible (too small) or take up the entire screen_. I think actually min/max canvas size is useful _generally_, ie it might be good to turn it on by default.

Fair enough, color me convinced!

Here's the current result with clamping:


https://user-images.githubusercontent.com/23482191/145976206-9e763da7-3d29-4c1b-9e79-481b3cc47bb9.mp4

As you can see, the minimum size is broken; this is caused by antialiasing, somehow. There is some place where aliasing gets multiplied by the size in the original shader, and things get essentially made transparent at very small sizes. This is twice as bad when the edge_width is not zero. Here's without edge:

https://user-images.githubusercontent.com/23482191/145976655-e1d00b12-1e11-431e-b993-9d782ac62a4a.mp4

You can kinda tell that the point is disappearing from the outside in. Not sure if there's a straightforward way to solve this other than using the `_antialias` property from #3430 to disable it. Here's with antialiasing disabled:


https://user-images.githubusercontent.com/23482191/145977055-9780067c-08cd-4ec9-9b88-c4080ffb9142.mp4

**Comment by jni** [comment]
I love this @brisvag! And I particularly love `canvas_size_limits` as an API!

Perhaps we have `_antialias` automatically return `False` if `canvas_size_limits[0] > 0`? ie:

```python
@property
def _antialias(self):
    return self.__antialias and self.experimental_canvas_size_limits[0] == 0
```

?

**Comment by brisvag** [comment]
Only the getter? The more these things come up, and the more I feel we should simply disable antialias by default until things are fixed on the vispy side...

**Comment by berl** [comment]
@jni  the `experimental` tag is great with me- I'll check this out

**Comment by brisvag** [comment]
[This test is annoying](https://github.com/napari/napari/runs/4609239021?check_suite_focus=true#step:9:652). @tlambert03 , `git` `blame`s you :P Thoughts?

**Comment by tlambert03** [comment]
that's not a test i wrote, just touched the decorator recently.  In general, I was trying to remove pytest-timeout altogether from the tests, exactly because it fails sometimes very confusingly like this.  see discussion in https://github.com/napari/napari/pull/3705

Basically, I believe the goal of that test is just to make sure that a dask array doesn't get greedily computed.  It was a very early way to test that (basically, start up a thread, and if it times out  in a couple seconds before calc_data_range is done, assume it failed).  But if it fails, these days it's usually because of the test approach, not the underlying thing (as the comment above it indicates).  I would not actually be opposed to removing that test altogether (since it's tested in other ways too).  But @sofroniewn  would need to weigh in on that

**Comment by sofroniewn** [comment]
> Basically, I believe the goal of that test is just to make sure that a dask array doesn't get greedily computed. It was a very early way to test that (basically, start up a thread, and if it times out in a couple seconds before calc_data_range is done, assume it failed). But if it fails, these days it's usually because of the test approach, not the underlying thing (as the comment above it indicates). I would not actually be opposed to removing that test altogether (since it's tested in other ways too). But @sofroniewn would need to weigh in on that

Very valid points, I've created #3869 to continue discussion and hopefully resolve this

**Comment by brisvag** [comment]
Thanks, I rememebr ecnountering this issue a couple times before. Then, I think this is ready for review/merge, since the only failing test is that one!

File: napari/layers/points/points.py
Comment (sofroniewn) [inline napari/layers/points/points.py]: Can you maybe add a comment here or elsewhere that without this implemented selection of points won't always work - just so we don't forget. I could see someone trying that and getting confused/ not remembering discussion on PR

File: napari/layers/points/points.py
Comment (sofroniewn) [inline napari/layers/points/points.py:303]: Is the max size of `10000` significant? Can we do "np.inf"? or is that problematic for other reasons. I might do 2^16-1 or something really really big if we don't expect this to come into play.

**Review by sofroniewn — APPROVED** [review]
Looks great! Left some fairly minor comments. Would be great to get @jni, @kevinyamauchi or @alisterburt to give this a spin before merge too. Thanks!!

File: napari/layers/points/points.py
Comment (brisvag) [reply napari/layers/points/points.py:303]: I don't know how `np.inf` would play here. I suppose we can/should give a max float value? Not sure what that is in glsl floats though. Certainly we can go bigger than 10000 though :P

Comment (sofroniewn) [reply napari/layers/points/points.py:303]: if `2^16-1` works then i think that is fine

Comment (brisvag) [reply napari/layers/points/points.py:303]: Oh, I just remembered why I simply put 10000: `gl_PointSize` is anyways limited by hardware/driver, and 10000 is way bigger than [anything we might even get close to](https://github.com/vispy/vispy/issues/2078#issuecomment-864078398). Don't forget these are screen pixels!

File: napari/layers/points/points.py
Comment (brisvag) [reply napari/layers/points/points.py]: added some details!

File: napari/layers/points/points.py
Comment (sofroniewn) [reply napari/layers/points/points.py:303]: oooh ok, that makes more sense then! let's leave at 10k then. i forgot they were screen pixels

File: napari/layers/points/points.py
Comment (alisterburt) [inline napari/layers/points/points.py]:
```suggestion
    size : float, array
        Size of the point marker in data pixels. If given as a scalar, 
        all points are made the same size. If given as an array, 
        size must be the same or broadcastable to the same shape 
        as the data.
```

**Review by alisterburt — APPROVED** [review]
Looks good!! One minor docstring change but let's get this merged! Thanks @brisvag 🙂

File: napari/layers/points/points.py
Comment (brisvag) [reply napari/layers/points/points.py]: Done!

**Pull request merged**
